### PR TITLE
[New-Engine]Support Hdfs Storage and async storage

### PR DESCRIPTION
**Pull request by CalvinKirs**
https://github.com/apache/incubator-seatunnel/issues/2274 subtask
[New-Engine]Support Hdfs Storage and async storage
Support config storage namespace
Support windows local file storage
Support async submit the task

### CODE REVIEW

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java [hunk 1]
```diff
@@ -79,4 +96,73 @@ public byte[] serializeCheckPointData(PipelineState state) throws IOException {
...
+     *
+     * @param fileNames file names
+     * @return latest checkpoint file name
+     */
+    public String getLatestCheckpointFileNameByJobIdAndPipelineId(List<String> fileNames, String pipelineId) {
+        AtomicReference<String> latestFileName = new AtomicReference<>();
+        AtomicInteger latestVersion = new AtomicInteger();
+        fileNames.forEach(fileName -> {
+            int fileVersion = Integer.parseInt(fileName.split(FILE_NAME_SPLIT)[FILE_SORT_ID_INDEX]);
+            String filePipelineId = fileName.split(FILE_NAME_SPLIT)[FILE_NAME_PIPELINE_ID_INDEX];
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java]: `fileName` only need to be split once

example
```java
String[] fileNameSegments = fileName.split(FILE_NAME_SPLIT);
```

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java [hunk 2]
```diff
@@ -79,4 +96,73 @@ public byte[] serializeCheckPointData(PipelineState state) throws IOException {
...
+     * @return latest checkpoint file name
+     */
+    public String getPipelineIdByFileName(String fileName) {
+        return fileName.split(FILE_NAME_SPLIT)[FILE_NAME_PIPELINE_ID_INDEX];
+    }
+
+    @Override
+    public void asyncStoreCheckPoint(PipelineState state) {
+        if (null == this.executor || this.executor.isShutdown()) {
+            this.executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2 + 1, new StorageThreadFactory());
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java]: Does `coreSize` need to define a configuration variable?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/CheckpointStorage.java [hunk 3]
```diff
@@ -35,6 +35,14 @@ public interface CheckpointStorage {
      */
     String storeCheckPoint(PipelineState state) throws CheckpointStorageException;
 
+    /**
+     * async save checkpoint to storage
+     *
+     * @param state PipelineState
+     * @throws CheckpointStorageException if save checkpoint failed
+     */
+    void asyncStoreCheckPoint(PipelineState state) throws CheckpointStorageException;
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/CheckpointStorage.java:44]: Need to return `Future`?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/common/StorageThreadFactory.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ * under the License.
+ *
+ */
+
+package org.apache.seatunnel.engine.checkpoint.storage.common;
+
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class StorageThreadFactory implements ThreadFactory {
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/common/StorageThreadFactory.java:26]: Use guava `ThreadFactoryBuilder` create ThreadFactory?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+            datas = serializeCheckPointData(state);
+        } catch (IOException e) {
+            throw new CheckpointStorageException("Failed to serialize checkpoint data", e);
+        }
+        //Consider file paths for different operating systems
+        String fileName = getStorageParentDirectory() + state.getJobId() + "/" + getCheckPointName(state);
+        try (FSDataOutputStream out = fs.create(new Path(fileName))) {
+            out.write(datas);
+        } catch (IOException e) {
+            throw new CheckpointStorageException("Failed to write checkpoint data", e);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java]: Log fileName or add fileName into exception message?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        if (fileNames.isEmpty()) {
+            throw new CheckpointStorageException("No checkpoint found for job, job id is: " + jobId);
+        }
+        Set<String> latestPipelineNames = getLatestPipelineNames(fileNames);
+        List<PipelineState> latestPipelineStates = new ArrayList<>();
+        latestPipelineNames.forEach(fileName -> {
+            try {
+                latestPipelineStates.add(readPipelineState(fileName, jobId));
+            } catch (CheckpointStorageException e) {
+                log.error("Failed to read pipeline state for file: {}", fileName, e);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java:152]: Should throws exception?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        }
+
+        List<PipelineState> pipelineStates = new ArrayList<>();
+        fileNames.forEach(file -> {
+            String filePipelineId = file.split(FILE_NAME_SPLIT)[FILE_NAME_PIPELINE_ID_INDEX];
+            if (pipelineId.equals(filePipelineId)) {
+                try {
+                    pipelineStates.add(readPipelineState(file, jobId));
+                } catch (Exception e) {
+                    log.error("Failed to read checkpoint data from file " + file, e);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java:192]: Should throws exception?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        return pipelineStates;
+    }
+
+    @Override
+    public void deleteCheckpoint(String jobId) {
+        String jobPath = getStorageParentDirectory() + jobId;
+        try {
+            fs.delete(new Path(jobPath), true);
+        } catch (IOException e) {
+            log.error("Failed to delete checkpoint for job {}", jobId, e);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java:205]: Should throws exception?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-local-file/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/localfile/LocalFileStorage.java [hunk 9]
```diff
@@ -138,32 +156,32 @@ public List<PipelineState> getLatestCheckpoint(String jobId) throws CheckpointSt
...
-            if (pipelineId.equals(filePipelineId) && fileVersion > latestVersion.get()) {
-                latestVersion.set(fileVersion);
-                latestFile.set(file);
+            String fileName = file.getName();
+            if (fileName.equals(latestFileName)) {
+                try {
+                    byte[] data = FileUtils.readFileToByteArray(file);
+                    latestFile.set(deserializeCheckPointData(data));
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-local-file/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/localfile/LocalFileStorage.java]: Should throws exception?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+        List<String> fileNames = getFileNames(path);
+        if (fileNames.isEmpty()) {
+            throw new CheckpointStorageException("No checkpoint found for job, job id is: " + jobId);
+        }
+        List<PipelineState> states = new ArrayList<>();
+        fileNames.forEach(file -> {
+            try {
+                states.add(readPipelineState(file, jobId));
+            } catch (CheckpointStorageException e) {
+                log.error("Failed to read checkpoint data from file: " + file, e);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java:130]: Should throws exception?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-local-file/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/localfile/LocalFileStorage.java [hunk 11]
```diff
@@ -107,27 +132,20 @@ public List<PipelineState> getLatestCheckpoint(String jobId) throws CheckpointSt
...
-                int oldVersion = Integer.parseInt(latestPipelineMap.get(filePipelineId).getName().split(FILE_NAME_SPLIT)[FILE_SORT_ID_INDEX]);
-                if (fileVersion > oldVersion) {
-                    latestPipelineMap.put(filePipelineId, file);
+            String fileName = file.getName();
+            if (latestPipelines.contains(fileName)) {
+                try {
+                    byte[] data = FileUtils.readFileToByteArray(file);
+                    latestPipelineFiles.add(deserializeCheckPointData(data));
+                } catch (IOException e) {
+                    log.error("Failed to read checkpoint data from file " + file.getAbsolutePath(), e);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-local-file/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/localfile/LocalFileStorage.java:145]: Should throws exception?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/common/ProtoStuffSerializer.java [hunk 12]
```diff
@@ -48,12 +43,13 @@ private static <T> Schema<T> getSchema(Class<T> clazz) {
     @Override
     public <T> byte[] serialize(T obj) {
         Class<T> clazz = (Class<T>) obj.getClass();
+        LinkedBuffer buffer = LinkedBuffer.allocate(LinkedBuffer.DEFAULT_BUFFER_SIZE);
```
Comment (ashulin) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/common/ProtoStuffSerializer.java:46]: I don't quite understand the change here, could you please explain it?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java (hunk 5 above)
Comment (hailin0) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java]: Write to temporary file(xxx.tmp) first and then rename?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java (hunk 8 above)
Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java:205]: No,For me this is a step that needs to be logged but doesn't need to throw an exception.

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/common/ProtoStuffSerializer.java (hunk 12 above)
Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/common/ProtoStuffSerializer.java:46]: Just serialized buffers.he cannot share.

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/CheckpointStorage.java (hunk 3 above)
Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/CheckpointStorage.java:44]: No

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java (hunk 6 above)
Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java:152]: No. This method is what it is, if you read its interface documentation.

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/common/StorageThreadFactory.java (hunk 4 above)
Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/common/StorageThreadFactory.java:26]: For me, I prefer Java native.

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java (hunk 5 above)
Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java]: > Write to temporary file(xxx.tmp) first and then rename?

could you tell me why? I don't think it's good way.

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java (hunk 1 above)
Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java]: good catch.

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java (hunk 2 above)
Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java]: We currently do not have a unified Config acquisition. We  use `System.getProperties` get,WDYT?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java [hunk 13]
```diff
@@ -79,4 +96,75 @@ public byte[] serializeCheckPointData(PipelineState state) throws IOException {
...
+     *
+     * @param fileName file names. note: file name cannot contain parent path
+     * @return latest checkpoint file name
+     */
+    public String getPipelineIdByFileName(String fileName) {
+        return fileName.split(FILE_NAME_SPLIT)[FILE_NAME_PIPELINE_ID_INDEX];
+    }
+
+    @Override
+    public void asyncStoreCheckPoint(PipelineState state) {
```
Comment (Hisoka-X) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java:163]: Does this method have multi-thread access？If so, the executor maybe create not only one time.

Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java:163]: Good catch~

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java (hunk 2 above)
Comment (EricJoy2048) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java]: > We currently do not have a unified Config acquisition. We use `System.getProperties` get,WDYT?

Actually, we already have config , look at `EngineConfig` class

Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java]: It's great if so, I think we couldn't need `Map<String,String> config` I will update this in next PR

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java [hunk 14]
```diff
@@ -23,36 +23,58 @@
...
     private final Serializer serializer = new ProtoStuffSerializer();
 
+    public static final String DEFAULT_CHECKPOINT_FILE_PATH_SPLIT = "/";
+
     /**
      * storage root directory
+     * if not set, use default value
      */
-    private static final String CHECKPOINT_DEFAULT_FILE_DIR = "/tmp/seatunnel/checkpoint/";
+    private String storageNameSpace = "/seatunnel/checkpoint/";
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java:59]: Use `${seatunnel installation directory}/checkpoint` ?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+        } catch (IOException e) {
+            throw new CheckpointStorageException("Failed to serialize checkpoint data,state is :" + state, e);
+        }
+        Path filePath = new Path(getStorageParentDirectory() + state.getJobId() + "/" + getCheckPointName(state));
+
+        Path tmpFilePath = new Path(getStorageParentDirectory() + state.getJobId() + "/" + getCheckPointName(state) + STORAGE_TMP_SUFFIX);
+        try (FSDataOutputStream out = fs.create(tmpFilePath)) {
+            out.write(datas);
+            out.hsync();
+            out.hflush();
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java]: hsync includes hflush capabilities?

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java (hunk 14 above)
Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java:59]: This seems redundant, for filesystems, which are usually configured by the user, and if the default directory is used, the naming `${seatunnel installation directory}` will be meaningless,

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java (hunk 15 above)
Comment (CalvinKirs) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-plugins/checkpoint-storage-hdfs/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/hdfs/HdfsStorage.java]: thx~ you are right

File: seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java (hunk 14 above)
Comment (hailin0) [reply seatunnel-engine/seatunnel-engine-storage/checkpoint-storage-api/src/main/java/org/apache/seatunnel/engine/checkpoint/storage/api/AbstractCheckpointStorage.java:59]: 👌

**Review by Hisoka-X — APPROVED** [review]

**Pull request merged**